src/tests/test_upgrade_agent.py

```diff
@@ -0,0 +1,189 @@
+import datetime
+import json
+import os
+import re
+import subprocess
+from typing import List
+
+import kubernetes.client
+import kubernetes.config
+import kubernetes.dynamic
+from junit_report import JunitTestSuite
+
+from service_client import log
+from tests.base_test import BaseTest
+
+
+def skopeo(args: List[str]) -> str:
+    """
+    Runs the 'skopeo' command with the given command line arguments. Raises an exception if the command finishes with
+    a non zero exit code, or if it writes anything to the standard error stream. Returns an string containing whatever
+    the command wrote to the standard output stream.
+    """
+    process = subprocess.Popen(
+        args=["skopeo"] + args,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    (stdout, stderr) = process.communicate()
+    if process.returncode != 0:
+        raise Exception(f"failed to run skopeo with args '{args}', finished with code {process.returncode}")
+    if len(stderr) > 0:
+        raise Exception(f"failed to run skopeo with args '{args}', wrote '{stderr}' to the standard error stream")
+    return stdout.decode("utf-8")
+
+
+def image_repo(image: str) -> str:
+    """
+    Calculates the image repository from the given image reference. For example, if the image is
+    'quay.io/my/image:v1.23' it will return 'quay.io/my/image'.
+    """
+    return re.sub(r":.*$", "", image)
+
+
+def image_digest(image: str) -> str:
+    """
+    Gets the digest of the given image.
+    """
```

Comment:
nit:
```suggestion
    """Gets the digest of the given image."""
```

Reply:
same elsewhere where it's just a short line

Reply:
Done.